RecoParticleFlow/PFProducer/src/PFBlockAlgo.cc

```diff
@@ -119,8 +119,22 @@ void PFBlockAlgo::setLinkers(const std::vector<edm::ParameterSet>& confs) {
       kdtrees_.emplace_back(KDTreeLinkerFactory::get()->create(pfx_kdtree + linkerName));
       kdtrees_.back()->setTargetType(std::min(type1, type2));
       kdtrees_.back()->setFieldType(std::max(type1, type2));
-    }
-  }
+      // giving entrance and exit points - necessary for track-hcal links
+      const std::string trajectoryLayerEntranceString(conf.existsAs<std::string>("trajectoryLayerEntrance")
+                                                          ? conf.getParameter<std::string>("trajectoryLayerEntrance")
+                                                          : "");
+      const std::string trajectoryLayerExitString(conf.existsAs<std::string>("trajectoryLayerExit")
+                                                      ? conf.getParameter<std::string>("trajectoryLayerExit")
+                                                      : "");
+      // convert TrajectoryLayers info from string to enum
+      const reco::PFTrajectoryPoint::LayerType trajectoryLayerEntrance =
+          reco::PFTrajectoryPoint::layerTypeByName(trajectoryLayerEntranceString);
+      const reco::PFTrajectoryPoint::LayerType trajectoryLayerExit =
+          reco::PFTrajectoryPoint::layerTypeByName(trajectoryLayerExitString);
+      if (trajectoryLayerEntrance != reco::PFTrajectoryPoint::Unknown)
+        kdtrees_.back()->setTrajectoryPoints(trajectoryLayerEntrance, trajectoryLayerExit);
```

Comment:
I would just pass conf to the corresponding tree linker and let the concrete linker constructor implementation figure out what to do with the inputs.
This way the ```setTrajectoryPoints``` extension to the interface is not needed (as it stands, it is needed only for one case).